### PR TITLE
Fix #246

### LINKED ISSUE #246: Freeze upon loading Intel Tatlow images
UEFITool seems to (usually) freeze when trying to load [Intel Tatlow (RKL-E)](https://ark.intel.com/content/www/us/en/ark/products/series/213678/intel-c250-series-chipsets.html) SPI images. Sample links below:

https://www.supermicro.com/Bios/softfiles/14259/BIOS_X12STD-F-1C31_20210831_1.0_STDsp.zip
https://dlcdnets.asus.com/pub/ASUS/server/P12R-M/BIOS/P12R-M-ASUS-0211.zip
https://dlcdnets.asus.com/pub/ASUS/server/P12R-E_Series/BIOS&FW/P12R-E-ASUS-0301.zip

The latest commits do generate a lot of warnings when compiling but I believe this freeze was happening before as well.

```
UEFITool-new_engine\common\nvramparser.cpp: In member function 'USTATUS NvramParser::parseNvarStore(const QModelIndex&)':
UEFITool-new_engine\common\nvramparser.cpp:350:72: warning: unknown conversion type character 'l' in format [-Wformat=]
                 info += usprintf("\nTimestamp: %" PRIX64 "h", timestamp);
                                                                        ^
UEFITool-new_engine\common\nvramparser.cpp:350:72: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\nvramparser.cpp: In member function 'USTATUS NvramParser::findNextStore(const QModelIndex&, const QByteArray&, UINT32, UINT32, UINT32&)':
UEFITool-new_engine\common\nvramparser.cpp:632:195: warning: unknown conversion type character 'l' in format [-Wformat=]
                     msg(usprintf("%s: FTW block candidate at offset %Xh skipped, has invalid body size %llXh", __FUNCTION__, localOffset + offset, (unsigned long long)ftw64Header->WriteQueueSize), index);
                                                                                                                                                                                                   ^
UEFITool-new_engine\common\nvramparser.cpp:632:195: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\nvramparser.cpp: In member function 'USTATUS NvramParser::parseVssStoreBody(const QModelIndex&, UINT8)':
UEFITool-new_engine\common\nvramparser.cpp:1538:96: warning: unknown conversion type character 'l' in format [-Wformat=]
             info += usprintf("\nMonotonic counter: %" PRIX64 "h\nTimestamp: ", monotonicCounter) + efiTimeToUString(timestamp)
                                                                                                ^
UEFITool-new_engine\common\nvramparser.cpp:1538:96: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\nvramparser.cpp: In member function 'USTATUS NvramParser::parseFlashMapBody(const QModelIndex&)':
UEFITool-new_engine\common\nvramparser.cpp:1928:36: warning: unknown conversion type character 'l' in format [-Wformat=]
                 entryHeader->Offset);
                                    ^
UEFITool-new_engine\common\nvramparser.cpp:1928:36: warning: format '%X' expects argument of type 'unsigned int', but argument 4 has type 'UINT64 {aka long long unsigned int}' [-Wformat=]
UEFITool-new_engine\common\nvramparser.cpp:1928:36: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\meparser.cpp: In member function 'USTATUS MeParser::parseFptRegion(const QByteArray&, const QModelIndex&, QModelIndex&)':
UEFITool-new_engine\common\meparser.cpp:180:38: warning: format '%lX' expects argument of type 'long unsigned int', but argument 2 has type 'unsigned int' [-Wformat=]
                         ptEntry->Type);
                                      ^
UEFITool-new_engine\common\meparser.cpp:180:38: warning: format '%lu' expects argument of type 'long unsigned int', but argument 3 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\meparser.cpp: In member function 'USTATUS MeParser::parseIfwi16Region(const QByteArray&, const QModelIndex&, QModelIndex&)':
UEFITool-new_engine\common\meparser.cpp:336:69: warning: unknown conversion type character 'l' in format [-Wformat=]
                             (unsigned long long)ifwiHeader->Checksum);
                                                                     ^
UEFITool-new_engine\common\meparser.cpp:336:69: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\utility.cpp: In function 'QString errorCodeToUString(USTATUS)':
UEFITool-new_engine\common\utility.cpp:187:93: warning: format '%lX' expects argument of type 'long unsigned int', but argument 2 has type 'USTATUS {aka unsigned int}' [-Wformat=]
     default:                                return usprintf("Unknown error %02lX", errorCode);
                                                                                             ^
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parsePeImageSectionBody(const QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:3167:42: warning: unknown conversion type character 'l' in format [-Wformat=]
             optionalHeader.H64->ImageBase);
                                          ^
UEFITool-new_engine\common\ffsparser.cpp:3167:42: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseTeImageSectionBody(const QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:3209:87: warning: unknown conversion type character 'l' in format [-Wformat=]
             teHeader->ImageBase + teHeader->StrippedSize - sizeof(EFI_IMAGE_TE_HEADER));
                                                                                       ^
UEFITool-new_engine\common\ffsparser.cpp:3209:87: warning: unknown conversion type character 'l' in format [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:3209:87: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::addInfoRecursive(const QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:3377:115: warning: unknown conversion type character 'l' in format [-Wformat=]
                 model->addInfo(index, usprintf("Data address: %08llXh\n", (unsigned long long)address + headerSize),false);
                                                                                                                   ^
UEFITool-new_engine\common\ffsparser.cpp:3377:115: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\ffsparser.cpp:3378:104: warning: unknown conversion type character 'l' in format [-Wformat=]
                 model->addInfo(index, usprintf("Header address: %08llXh\n", (unsigned long long)address), false);
                                                                                                        ^
UEFITool-new_engine\common\ffsparser.cpp:3378:104: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\ffsparser.cpp:3381:97: warning: unknown conversion type character 'l' in format [-Wformat=]
                 model->addInfo(index, usprintf("Address: %08llXh\n", (unsigned long long)address), false);
                                                                                                 ^
UEFITool-new_engine\common\ffsparser.cpp:3381:97: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseFit(const QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:3862:83: warning: unknown conversion type character 'l' in format [-Wformat=]
         currentStrings.push_back(usprintf("%016" PRIX64 "h", currentEntry->Address));
                                                                                   ^
UEFITool-new_engine\common\ffsparser.cpp:3862:83: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseFitEntryAcm(const QByteArray&, UINT32, const QModelIndex&, QString&, UINT32&)':
UEFITool-new_engine\common\ffsparser.cpp:4013:25: warning: format '%lX' expects argument of type 'long unsigned int', but argument 5 has type 'unsigned int' [-Wformat=]
                         );
                         ^
UEFITool-new_engine\common\ffsparser.cpp:4013:25: warning: format '%lX' expects argument of type 'long unsigned int', but argument 13 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:4013:25: warning: format '%lX' expects argument of type 'long unsigned int', but argument 21 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:4013:25: warning: format '%lX' expects argument of type 'long unsigned int', but argument 22 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseFitEntryBootGuardBootPolicy(const QByteArray&, UINT32, const QModelIndex&, QString&, UINT32&)':
UEFITool-new_engine\common\ffsparser.cpp:4208:39: warning: unknown conversion type character 'l' in format [-Wformat=]
                                       );
                                       ^
UEFITool-new_engine\common\ffsparser.cpp:4208:39: warning: unknown conversion type character 'l' in format [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:4208:39: warning: format '%X' expects argument of type 'unsigned int', but argument 6 has type 'UINT64 {aka long long unsigned int}' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:4208:39: warning: format '%X' expects argument of type 'unsigned int', but argument 7 has type 'UINT64 {aka long long unsigned int}' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:4208:39: warning: too many arguments for format [-Wformat-extra-args]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseBpdtRegion(const QByteArray&, UINT32, UINT32, const QModelIndex&, QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:4562:38: warning: format '%lX' expects argument of type 'long unsigned int', but argument 2 has type 'unsigned int' [-Wformat=]
                         ptEntry->Size) +
                                      ^
UEFITool-new_engine\common\ffsparser.cpp:4562:38: warning: format '%lu' expects argument of type 'long unsigned int', but argument 3 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseCpdRegion(const QByteArray&, UINT32, const QModelIndex&, QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lX' expects argument of type 'long unsigned int', but argument 3 has type 'unsigned int' [-Wformat=]
                                          manifestHeader->ExponentSize * sizeof(UINT32), manifestHeader->ExponentSize * sizeof(UINT32));
                                                                                                                                      ^
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lX' expects argument of type 'long unsigned int', but argument 9 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lu' expects argument of type 'long unsigned int', but argument 10 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lX' expects argument of type 'long unsigned int', but argument 16 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lu' expects argument of type 'long unsigned int', but argument 17 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lX' expects argument of type 'long unsigned int', but argument 18 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp:5015:134: warning: format '%lu' expects argument of type 'long unsigned int', but argument 19 has type 'unsigned int' [-Wformat=]
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseCpdExtensionsArea(const QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:5176:173: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'unsigned int' [-Wformat=]
                     msg(usprintf("%s: IFWI Partition Manifest hash size is %d, expected %lu", __FUNCTION__, attrHeader->HashSize, sizeof (attrHeader->CompletePartitionHash)), extIndex);
                                                                                                                                                                             ^
UEFITool-new_engine\common\ffsparser.cpp:5202:143: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'unsigned int' [-Wformat=]
                     msg(usprintf("%s: Module Attributes hash size is %d, expected %lu", __FUNCTION__, hashSize, sizeof (attrHeader->ImageHash)), extIndex);
                                                                                                                                               ^
UEFITool-new_engine\common\ffsparser.cpp: In member function 'USTATUS FfsParser::parseSignedPackageInfoData(const QModelIndex&)':
UEFITool-new_engine\common\ffsparser.cpp:5253:164: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'unsigned int' [-Wformat=]
                 msg(usprintf("%s: CPD Signed Package Info hash size is %d, expected %lu", __FUNCTION__, moduleHeader->HashSize, sizeof (moduleHeader->MetadataHash)), extIndex);
```

### PR DESCRIPTION
**Pull request by NikolajSchlej**
- fixed CPLD extension area parser hang
- added some definitions for CSME types obtained from MEParser

### CODE REVIEW

File: common/ffsparser.cpp [hunk 1]
```diff
@@ -5106,7 +5089,7 @@ USTATUS FfsParser::parseCpdExtensionsArea(const UModelIndex & index)
     UINT32 offset = 0;
     while (offset < (UINT32)body.size()) {
         const CPD_EXTENTION_HEADER* extHeader = (const CPD_EXTENTION_HEADER*) (body.constData() + offset);
-        if (extHeader->Length <= ((UINT32)body.size() - offset)) {
```
Comment (NikolajSchlej) [inline common/ffsparser.cpp:5109]: This is what fixes the hang, but it's still not perfect. All this shit had been written in hurry for a specific case and begs for a complete rewrite.

File: common/ffs.h [hunk 2]
```diff
@@ -740,15 +741,19 @@ typedef struct CPD_EXTENTION_HEADER_ {
...
+#define CPD_EXT_TYPE_SPS_PLATFORM_ID         50
 
 typedef struct CPD_EXT_SIGNED_PACKAGE_INFO_MODULE_ {
     UINT8  Name[12];
     UINT8  Type;
     UINT8  HashAlgorithm;
     UINT16 HashSize;
     UINT32 MetadataSize;
-    UINT8  MetadataHash[32];
+    UINT8  MetadataHash[1]; // Can be 32 or 48 bit
```
Comment (vit9696) [inline common/ffs.h:756]: Use `MetadataHash[];`. Even MSVC supports this today, and GCC can optimise out any access to this array with indices larger than 0 anyway.

File: common/ffs.h [hunk 3]
```diff
@@ -765,12 +770,11 @@ typedef struct CPD_EXT_SIGNED_PACKAGE_INFO_ {
...
     UINT32 ExtensionType;
     UINT32 ExtensionLength;
-    UINT8  CompressionType;
-    UINT8  Reserved[3];
+    UINT32 CompressionType;
     UINT32 UncompressedSize;
     UINT32 CompressedSize;
     UINT32 GlobalModuleId;
-    UINT8  ImageHash[32];
+    UINT8  ImageHash[1]; // The actual hash size is 32 or 48 bytes
```
Comment (vit9696) [inline common/ffs.h:777]: Same here.

File: common/ffsparser.cpp (hunk 1 above)
Comment (vit9696) [reply common/ffsparser.cpp:5109]: Yes… but should do for the time being.

File: common/ffs.h (hunk 2 above)
Comment (NikolajSchlej) [reply common/ffs.h:756]: I'm really unsure it's supported in anything that can still target Windows XP, and I still want to.

Comment (vit9696) [reply common/ffs.h]: Yeah, rewrote them with sizeof offsets, looks good too.

**Pull request merged**